File: paleo_workbench/mapping/map_styles.py
"""Unified cartographic style and symbol definitions for map rendering.

Styles are pure data (no Qt imports) so the authoring document, the render
snapshot seam, persistence, and both render backends share one vocabulary.
``VectorStyle.to_dict`` keeps the established flat dict keys (``fill``,
``stroke``, ``stroke_width``, ``marker_size``, ``labels``) and only adds new
ones, so persisted projects and the QGIS bridge payload stay compatible.
"""

from __future__ import annotations

from dataclasses import dataclass, field, replace
from enum import Enum
import json
from pathlib import Path
from typing import Any, Mapping

__all__ = [
    "LinePattern",
    "MarkerSymbol",
    "TextStyle",
    "VectorStyle",
    "STYLE_LIBRARY",
    "default_style_for",
    "load_style_library",
    "save_style_library",
    "style_dict_revision",
]


class LinePattern(str, Enum):
    """Named cartographic line patterns for geological map features."""

    SOLID = "solid"
    DASH = "dash"
    DOT = "dot"
    DASH_DOT = "dash_dot"
    # 断层线: the classic long-dash fault trace used on paleogeographic maps.
    FAULT = "fault"
    # 地层界线: solid with a heavier weight is expressed via stroke_width;
    # the pattern itself is solid, listed for symbol-library completeness.
    BOUNDARY = "boundary"

    def dash_pattern(self, width: float) -> tuple[float, ...]:
        """Qt dash units (multiples of pen width) for this pattern."""
        if self is LinePattern.DASH:
            return (4.0, 2.0)
        if self is LinePattern.DOT:
            return (1.0, 2.0)
        if self is LinePattern.DASH_DOT:
            return (4.0, 2.0, 1.0, 2.0)
        if self is LinePattern.FAULT:
            return (6.0, 2.0)
        return ()


class MarkerSymbol(str, Enum):
    """Point symbol vocabulary including standard well markers."""

    CIRCLE = "circle"
    SQUARE = "square"
    TRIANGLE = "triangle"
    DIAMOND = "diamond"
    CROSS = "cross"
    STAR = "star"
    # 井符号: ring with a centre dot, the standard well location mark.
    WELL = "well"


@dataclass(frozen=True, slots=True)
class TextStyle:
    """Label placement style consumed by labeling-capable backends."""

    field: str = ""
    size: float = 9.0
    color: str = "#f8f9fa"
    font_family: str = ""
    bold: bool = False
    halo_color: str = "#182431"
    halo_width: float = 1.0
    visible: bool = True
    # #1052: per-feature data-defined overrides honoured by the QGIS PAL
    # backend — attribute FIELD names (rotation in degrees clockwise, size
    # in points, colour as a colour string); "" disables each override and
    # the fixed values above apply.
    rotation_field: str = ""
    size_field: str = ""
    color_field: str = ""
    # #1102: explicit buffer (halo) colour. "" falls back to halo_color on
    # the QGIS wire; the native bridge defaults to white when neither is set.
    buffer_color: str = ""

    def to_dict(self) -> dict[str, Any]:
        return {
            "field": self.field,
            "size": self.size,
            "color": self.color,
            "font_family": self.font_family,
            "bold": self.bold,
            "halo_color": self.halo_color,
            "halo_width": self.halo_width,
            "visible": self.visible,
            "rotation_field": self.rotation_field,
            "size_field": self.size_field,
            "color_field": self.color_field,
            "buffer_color": self.buffer_color,
        }

    @classmethod
    def from_dict(cls, data: Mapping[str, Any] | None) -> "TextStyle":
        if not isinstance(data, Mapping):
            return cls()
        values = dict(data)
        style = cls()
        replacements: dict[str, Any] = {}
        for key in (
            "field",
            "color",
            "font_family",
            "halo_color",
            "rotation_field",
            "size_field",
            "color_field",
            "buffer_color",
        ):
            if values.get(key) is not None:
                replacements[key] = str(values[key])
        for key in ("size", "halo_width"):
            if values.get(key) is not None:
                try:
                    replacements[key] = float(values[key])
                except (TypeError, ValueError):
                    continue
        if values.get("bold") is not None:
            replacements["bold"] = bool(values["bold"])
        if values.get("visible") is not None:
            replacements["visible"] = bool(values["visible"])
        return replace(style, **replacements) if replacements else style


@dataclass(frozen=True, slots=True)
class VectorStyle:
    """One layer-level vector style with per-feature renderer settings.

    Stroke width, marker size and label size are logical pixels at 96 dpi;
    renderers scale them by ``dpi / 96`` so screen and export stay identical.
    """

    fill: str = "#6c8ebf"
    stroke: str = "#26364d"
    stroke_width: float = 1.0
    line_pattern: LinePattern = LinePattern.SOLID
    marker: MarkerSymbol = MarkerSymbol.CIRCLE
    marker_size: float = 6.0
    # Renderer classification (currently honoured by the QGIS backend; kept
    # here so one schema describes both backends).
    renderer: str = "single"
    field: str = ""
    categories: tuple[tuple[str, str, str], ...] = ()  # (value, fill, label)
    ranges: tuple[tuple[float, float, str, str], ...] = ()  # (lo, hi, fill, label)
    labels: TextStyle | None = None

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {
            "fill": self.fill,
            "stroke": self.stroke,
            "stroke_width": self.stroke_width,
            "line_pattern": self.line_pattern.value,
            "marker": self.marker.value,
            "marker_size": self.marker_size,
            "renderer": self.renderer,
            "field": self.field,
        }
        if self.categories:
            data["categories"] = [list(entry) for entry in self.categories]
        if self.ranges:
            data["ranges"] = [list(entry) for entry in self.ranges]
        if self.labels is not None:
            data["labels"] = self.labels.to_dict()
        return data

    @classmethod
    def from_dict(cls, data: Mapping[str, Any] | None) -> "VectorStyle":
        """Parse persisted/host style dicts tolerantly (unknown keys ignored)."""
        if not isinstance(data, Mapping):
            return cls()
        values = dict(data)
        replacements: dict[str, Any] = {}
        for key in ("fill", "stroke"):
            if values.get(key):
                replacements[key] = str(values[key])
        for key in ("stroke_width", "marker_size"):
            if values.get(key) is not None:
                try:
                    replacements[key] = max(0.0, float(values[key]))
                except (TypeError, ValueError):
                    continue
        if values.get("line_pattern"):
            try:
                replacements["line_pattern"] = LinePattern(str(values["line_pattern"]))
            except ValueError:
                pass
        if values.get("marker"):
            try:
                replacements["marker"] = MarkerSymbol(str(values["marker"]))
            except ValueError:
                pass
        for key in ("renderer", "field"):
            if values.get(key) is not None:
                replacements[key] = str(values[key])
        categories: list[tuple[str, str, str]] = []
        raw_categories = values.get("categories")
        if isinstance(raw_categories, Mapping):
            # Established QGIS payload form: {"value": "#color"}.
            categories.extend(
                (str(key), str(color), "") for key, color in raw_categories.items()
            )
        for entry in raw_categories if isinstance(raw_categories, (list, tuple)) else ():
            if isinstance(entry, (list, tuple)) and len(entry) >= 2:
                categories.append((str(entry[0]), str(entry[1]), str(entry[2]) if len(entry) > 2 else ""))
        if categories:
            replacements["categories"] = tuple(categories)
        ranges: list[tuple[float, float, str, str]] = []
        for entry in values.get("ranges") or ():
            if isinstance(entry, (list, tuple)) and len(entry) >= 3:
                try:
                    ranges.append(
                        (
                            float(entry[0]),
                            float(entry[1]),
                            str(entry[2]),
                            str(entry[3]) if len(entry) > 3 else "",
                        )
                    )
                except (TypeError, ValueError):
                    continue
            elif isinstance(entry, Mapping):
                try:
                    lo = float(entry.get("min", entry.get("lo", 0.0)))
                    hi = float(entry.get("max", entry.get("hi", 1.0)))
                    fill = str(entry.get("fill", entry.get("color", "#6c8ebf")))
                    lbl = str(entry.get("label", ""))
                    ranges.append((lo, hi, fill, lbl))
                except (TypeError, ValueError):
                    continue
        if ranges:
            replacements["ranges"] = tuple(ranges)
        if values.get("labels") is not None:
            replacements["labels"] = TextStyle.from_dict(values.get("labels"))
        return cls(**replacements) if replacements else cls()


# 断层/等值线/地层界线/井符号/注记/相带 named presets. Values intentionally
# match the previously hard-coded defaults so existing projects look unchanged.
STYLE_LIBRARY: dict[str, VectorStyle] = {
    "facies": VectorStyle(
        fill="#6c8ebf", stroke="#26364d", stroke_width=1.0,
    ),
    "well": VectorStyle(
        fill="#22b8a7", stroke="#182431", marker=MarkerSymbol.WELL, marker_size=7.0,
    ),
    "contour": VectorStyle(
        fill="transparent", stroke="#f08c46", stroke_width=1.0,
        labels=TextStyle(field="", size=8.0, color="#ffd8a8"),
    ),
    "formation_boundary": VectorStyle(
        fill="transparent", stroke="#e8590c", stroke_width=2.0, line_pattern=LinePattern.SOLID,
    ),
    "fault": VectorStyle(
        fill="transparent", stroke="#e03131", stroke_width=2.0, line_pattern=LinePattern.FAULT,
    ),
    "line": VectorStyle(
        fill="transparent", stroke="#f08c46", stroke_width=2.0,
    ),
    "annotation": VectorStyle(
        fill="#eff3f8", stroke="#182431", marker=MarkerSymbol.CIRCLE, marker_size=4.0,
        labels=TextStyle(
            field="text", size=10.0, color="#f8f9fa",
            # #1052: annotation features carry per-feature `rotation`,
            # `font_size`, and `color` properties (AnnotationMapLayer
            # ._sync_features_from_annotations); binding them here lets the
            # QGIS PAL backend honour each annotation's own angle/size/
            # colour instead of flattening every label to the fixed format.
            rotation_field="rotation",
            size_field="font_size",
            color_field="color",
        ),
    ),
    "label": VectorStyle(
        fill="#eff3f8", stroke="#182431", marker=MarkerSymbol.CIRCLE, marker_size=4.0,
    ),
}

_STYLE_FOR_KIND = {
    "facies": "facies",
    "well": "well",
    "line": "line",
    "label": "label",
    # #1052: AnnotationMapLayer.__post_init__ resolves its preset through
    # this map — without the entry it silently fell back to the facies
    # preset (labels=None), so annotation labels never had a default style
    # or the per-feature data-defined field bindings.
    "annotation": "annotation",
}


def default_style_for(kind: str) -> VectorStyle:
    """Return the symbol-library preset for a compatibility layer kind."""
    return STYLE_LIBRARY[_STYLE_FOR_KIND.get(str(kind), "facies")]


def save_style_library(path: Path | str, *, styles: Mapping[str, VectorStyle] | None = None) -> None:
    """Persist named styles as a JSON file for reuse across documents."""
    payload = {
        "schema_version": 1,
        "styles": {
            name: style.to_dict() for name, style in (styles or STYLE_LIBRARY).items()
        },
    }
    Path(path).write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")


def load_style_library(path: Path | str) -> dict[str, VectorStyle]:
    """Load a previously saved style library; unknown entries are ignored."""
    data = json.loads(Path(path).read_text(encoding="utf-8"))
    styles: dict[str, VectorStyle] = {}
    for name, entry in (data.get("styles") or {}).items():
        if isinstance(entry, Mapping):
            styles[str(name)] = VectorStyle.from_dict(entry)
    return styles


def style_dict_revision(style: Mapping[str, Any] | None) -> int:
    """Cheap stable revision for a style dict.

    Uses recursive tuple conversion (C-speed for str/float/tuple) instead of a
    JSON round-trip; style dicts are small so this stays well under a
    microsecond while remaining content-stable.
    """
    def freeze(value: Any) -> Any:
        if isinstance(value, Mapping):
            return tuple(sorted((str(key), freeze(item)) for key, item in value.items()))
        if isinstance(value, (list, tuple)):
            return tuple(freeze(item) for item in value)
        if isinstance(value, bool) or value is None:
            return value
        if isinstance(value, (int, float, str)):
            return value
        return str(value)

    return hash(freeze(style or {}))

File: libs/ui_composite/src/map_styles.cpp
#include <pwb/ui_composite/map_styles.hpp>

#include <algorithm>

namespace pwb::ui_composite {
namespace {

const Json* find(const Json& object, const char* key) {
    if (!object.is_object()) {
        return nullptr;
    }
    auto it = object.find(key);
    return it == object.end() ? nullptr : &*it;
}

std::string str_or(const Json& object, const char* key,
                   std::string fallback = {}) {
    const Json* value = find(object, key);
    if (value == nullptr || value->is_null()) {
        return fallback;
    }
    if (value->is_string()) {
        return value->get<std::string>();
    }
    if (value->is_number_integer()) {
        return std::to_string(value->get<long long>());
    }
    if (value->is_number()) {
        return std::to_string(value->get<double>());
    }
    if (value->is_boolean()) {
        return value->get<bool>() ? "True" : "False";
    }
    return fallback;
}

bool bool_or(const Json& object, const char* key, bool fallback) {
    const Json* value = find(object, key);
    if (value != nullptr && value->is_boolean()) {
        return value->get<bool>();
    }
    return fallback;
}

double num_or(const Json& object, const char* key, double fallback,
              double floor = -1e300) {
    const Json* value = find(object, key);
    if (value != nullptr && value->is_number()) {
        return std::max(floor, value->get<double>());
    }
    return fallback;
}

std::string known_line_pattern(const std::string& value,
                               const std::string& fallback) {
    static const char* kPatterns[] = {
        line_pattern::kSolid, line_pattern::kDash, line_pattern::kDot,
        line_pattern::kDashDot, line_pattern::kFault, line_pattern::kBoundary,
    };
    for (const char* pattern : kPatterns) {
        if (value == pattern) {
            return value;
        }
    }
    return fallback;
}

std::string known_marker(const std::string& value,
                         const std::string& fallback) {
    static const char* kMarkers[] = {
        marker_symbol::kCircle, marker_symbol::kSquare,
        marker_symbol::kTriangle, marker_symbol::kDiamond,
        marker_symbol::kCross,  marker_symbol::kStar,
        marker_symbol::kWell,
    };
    for (const char* marker : kMarkers) {
        if (value == marker) {
            return value;
        }
    }
    return fallback;
}

}  // namespace

std::vector<double> line_pattern_dash_pattern(const std::string& pattern,
                                              double width) {
    if (pattern == line_pattern::kDash) {
        return {4.0 * width, 2.0 * width};
    }
    if (pattern == line_pattern::kDot) {
        return {1.0 * width, 2.0 * width};
    }
    if (pattern == line_pattern::kDashDot) {
        return {4.0 * width, 2.0 * width, 1.0 * width, 2.0 * width};
    }
    if (pattern == line_pattern::kFault) {
        return {6.0 * width, 2.0 * width};
    }
    return {};
}

Json TextStyle::to_dict() const {
    return Json{
        {"field", field},
        {"size", size},
        {"color", color},
        {"font_family", font_family},
        {"bold", bold},
        {"halo_color", halo_color},
        {"halo_width", halo_width},
        {"visible", visible},
        {"rotation_field", rotation_field},
        {"size_field", size_field},
        {"color_field", color_field},
        {"buffer_color", buffer_color},
    };
}

TextStyle TextStyle::from_dict(const Json& data) {
    TextStyle style;
    if (!data.is_object()) {
        return style;
    }
    style.field = str_or(data, "field", style.field);
    style.color = str_or(data, "color", style.color);
    style.font_family = str_or(data, "font_family", style.font_family);
    style.halo_color = str_or(data, "halo_color", style.halo_color);
    style.rotation_field =
        str_or(data, "rotation_field", style.rotation_field);
    style.size_field = str_or(data, "size_field", style.size_field);
    style.color_field = str_or(data, "color_field", style.color_field);
    style.buffer_color = str_or(data, "buffer_color", style.buffer_color);
    style.size = num_or(data, "size", style.size);
    style.halo_width = num_or(data, "halo_width", style.halo_width);
    style.bold = bool_or(data, "bold", style.bold);
    style.visible = bool_or(data, "visible", style.visible);
    return style;
}

Json VectorStyle::to_dict() const {
    Json data = {
        {"fill", fill},
        {"stroke", stroke},
        {"stroke_width", stroke_width},
        {"line_pattern", line_pattern},
        {"marker", marker},
        {"marker_size", marker_size},
        {"renderer", renderer},
        {"field", field},
    };
    if (!categories.empty()) {
        Json entries = Json::array();
        for (const auto& [value, fill_color, label] : categories) {
            entries.push_back({value, fill_color, label});
        }
        data["categories"] = std::move(entries);
    }
    if (!ranges.empty()) {
        Json entries = Json::array();
        for (const auto& [lo, hi, fill_color, label] : ranges) {
            entries.push_back({lo, hi, fill_color, label});
        }
        data["ranges"] = std::move(entries);
    }
    if (!fill_patterns.empty()) {
        Json patterns = Json::object();
        for (const auto& [value, pattern_id] : fill_patterns) {
            patterns[value] = pattern_id;
        }
        data["fill_patterns"] = std::move(patterns);
    }
    if (labels.has_value()) {
        data["labels"] = labels->to_dict();
    }
    return data;
}

VectorStyle VectorStyle::from_dict(const Json& data) {
    VectorStyle style;
    if (!data.is_object()) {
        return style;
    }
    style.fill = str_or(data, "fill", style.fill);
    style.stroke = str_or(data, "stroke", style.stroke);
    style.stroke_width = num_or(data, "stroke_width", style.stroke_width, 0.0);
    style.marker_size = num_or(data, "marker_size", style.marker_size, 0.0);
    style.line_pattern =
        known_line_pattern(str_or(data, "line_pattern"), style.line_pattern);
    style.marker = known_marker(str_or(data, "marker"), style.marker);
    style.renderer = str_or(data, "renderer", style.renderer);
    style.field = str_or(data, "field", style.field);

    const Json* categories = find(data, "categories");
    if (categories != nullptr && categories->is_object()) {
        for (const auto& [key, value] : categories->items()) {
            style.categories.emplace_back(key,
                                          value.is_string()
                                              ? value.get<std::string>()
                                              : std::string{},
                                          std::string{});
        }
    } else if (categories != nullptr && categories->is_array()) {
        for (const Json& entry : *categories) {
            if (entry.is_array() && entry.size() >= 2) {
                style.categories.emplace_back(
                    entry[0].is_string() ? entry[0].get<std::string>()
                                         : std::string{},
                    entry[1].is_string() ? entry[1].get<std::string>()
                                         : std::string{},
                    entry.size() > 2 && entry[2].is_string()
                        ? entry[2].get<std::string>()
                        : std::string{});
            }
        }
    }

    const Json* ranges = find(data, "ranges");
    if (ranges != nullptr && ranges->is_array()) {
        for (const Json& entry : *ranges) {
            if (entry.is_array() && entry.size() >= 3) {
                if (entry[0].is_number() && entry[1].is_number()) {
                    style.ranges.emplace_back(
                        entry[0].get<double>(), entry[1].get<double>(),
                        entry[2].is_string() ? entry[2].get<std::string>()
                                             : std::string{},
                        entry.size() > 3 && entry[3].is_string()
                            ? entry[3].get<std::string>()
                            : std::string{});
                }
            } else if (entry.is_object()) {
                const Json* lo = find(entry, "min");
                if (lo == nullptr) {
                    lo = find(entry, "lo");
                }
                const Json* hi = find(entry, "max");
                if (hi == nullptr) {
                    hi = find(entry, "hi");
                }
                if (lo != nullptr && hi != nullptr && lo->is_number() &&
                    hi->is_number()) {
                    style.ranges.emplace_back(
                        lo->get<double>(), hi->get<double>(),
                        str_or(entry, "fill",
                               str_or(entry, "color", "#6c8ebf")),
                        str_or(entry, "label"));
                }
            }
        }
    }

    const Json* patterns = find(data, "fill_patterns");
    if (patterns != nullptr && patterns->is_object()) {
        for (const auto& [key, value] : patterns->items()) {
            style.fill_patterns.emplace_back(
                key, value.is_string() ? value.get<std::string>()
                                       : std::string{});
        }
    } else if (patterns != nullptr && patterns->is_array()) {
        for (const Json& entry : *patterns) {
            if (entry.is_array() && entry.size() >= 2 &&
                entry[0].is_string() && entry[1].is_string()) {
                style.fill_patterns.emplace_back(
                    entry[0].get<std::string>(), entry[1].get<std::string>());
            }
        }
    }

    const Json* labels = find(data, "labels");
    if (labels != nullptr && !labels->is_null()) {
        style.labels = TextStyle::from_dict(*labels);
    }
    return style;
}

const std::map<std::string, VectorStyle>& style_library() {
    static const std::map<std::string, VectorStyle> library = [] {
        std::map<std::string, VectorStyle> map;
        {
            VectorStyle facies;
            facies.fill = "#6c8ebf";
            facies.stroke = "#26364d";
            facies.stroke_width = 1.0;
            facies.labels = TextStyle{.field = "facies",
                                      .size = 9.0,
                                      .color = "#1f2937",
                                      .halo_color = "#f8f9fa",
                                      .halo_width = 1.0};
            map.emplace("facies", facies);
        }
        {
            VectorStyle well;
            well.fill = "#22b8a7";
            well.stroke = "#182431";
            well.marker = marker_symbol::kWell;
            well.marker_size = 7.0;
            well.labels = TextStyle{.field = "name",
                                    .size = 9.0,
                                    .color = "#1f2937",
                                    .halo_color = "#f8f9fa",
                                    .halo_width = 1.0};
            map.emplace("well", well);
        }
        {
            VectorStyle contour;
            contour.fill = "transparent";
            contour.stroke = "#f08c46";
            contour.stroke_width = 1.0;
            contour.labels = TextStyle{.field = "",
                                       .size = 8.0,
                                       .color = "#ffd8a8"};
            map.emplace("contour", contour);
        }
        {
            VectorStyle boundary;
            boundary.fill = "transparent";
            boundary.stroke = "#e8590c";
            boundary.stroke_width = 2.0;
            boundary.labels = TextStyle{.field = "name",
                                        .size = 9.0,
                                        .color = "#e8590c",
                                        .halo_color = "#f8f9fa",
                                        .halo_width = 1.0};
            map.emplace("formation_boundary", boundary);
        }
        {
            VectorStyle fault;
            fault.fill = "transparent";
            fault.stroke = "#e03131";
            fault.stroke_width = 2.0;
            fault.line_pattern = line_pattern::kFault;
            fault.labels = TextStyle{.field = "name",
                                     .size = 9.0,
                                     .color = "#c92a2a",
                                     .halo_color = "#f8f9fa",
                                     .halo_width = 1.0};
            map.emplace("fault", fault);
        }
        {
            VectorStyle line;
            line.fill = "transparent";
            line.stroke = "#f08c46";
            line.stroke_width = 2.0;
            map.emplace("line", line);
        }
        {
            VectorStyle annotation;
            annotation.fill = "#eff3f8";
            annotation.stroke = "#182431";
            annotation.marker = marker_symbol::kCircle;
            annotation.marker_size = 4.0;
            TextStyle labels;
            labels.field = "text";
            labels.size = 10.0;
            labels.color = "#f8f9fa";
            labels.rotation_field = "rotation";
            labels.size_field = "font_size";
            labels.color_field = "color";
            annotation.labels = labels;
            map.emplace("annotation", annotation);
        }
        {
            VectorStyle label;
            label.fill = "#eff3f8";
            label.stroke = "#182431";
            label.marker = marker_symbol::kCircle;
            label.marker_size = 4.0;
            map.emplace("label", label);
        }
        return map;
    }();
    return library;
}

const VectorStyle& default_style_for(const std::string& kind) {
    static const std::map<std::string, std::string> for_kind = {
        {"facies", "facies"},
        {"well", "well"},
        {"line", "line"},
        {"label", "label"},
        // #1052: annotation presets resolve through this map (labels +
        // per-feature data-defined bindings), never the facies fallback.
        {"annotation", "annotation"},
    };
    auto it = for_kind.find(kind);
    return style_library().at(it == for_kind.end() ? "facies" : it->second);
}

int64_t style_dict_revision(const Json& style) {
    // Cheap stable content hash — mirrors Python's recursive-tuple freeze:
    // dict keys sorted, sequences preserved, everything else by value.
    // std::hash composition is deterministic within a process; style dicts
    // are small so this stays well under a microsecond.
    struct Freeze {
        static void append(const Json& value, std::string& out) {
            if (value.is_object()) {
                std::vector<std::string> keys;
                keys.reserve(value.size());
                for (const auto& [key, item] : value.items()) {
                    keys.push_back(key);
                }
                std::sort(keys.begin(), keys.end());
                out += '{';
                for (const std::string& key : keys) {
                    out += key;
                    out += ':';
                    append(value.at(key), out);
                    out += ',';
                }
                out += '}';
                return;
            }
            if (value.is_array()) {
                out += '[';
                for (const Json& item : value) {
                    append(item, out);
                    out += ',';
                }
                out += ']';
                return;
            }
            out += value.dump();
        }
    };
    std::string frozen;
    if (!style.is_null()) {
        Freeze::append(style, frozen);
    }
    return static_cast<int64_t>(std::hash<std::string>{}(frozen));
}

}  // namespace pwb::ui_composite
